be/src/agent/publish_version.cpp

```diff
@@ -0,0 +1,152 @@
+                    LOG(WARNING) << task.st;
+                    return;
+                }
+                TabletSharedPtr tablet = StorageEngine::instance()->tablet_manager()->get_tablet(task.tablet_id);
+                if (!tablet) {
+                    task.st = Status::NotFound(
+                            fmt::format("Not found tablet to publish_version. tablet_id: {}, txn_id: {}",
+                                        task.tablet_id, task.txn_id));
+                    LOG(WARNING) << task.st;
+                    return;
+                }
+                {
+                    std::lock_guard lg(affected_dirs_lock);
+                    affected_dirs.insert(tablet->data_dir());
+                }
+                task.st = StorageEngine::instance()->txn_manager()->publish_txn(task.partition_id, tablet, task.txn_id,
+                                                                                task.version, task.rowset);
+                if (!task.st.ok()) {
+                    LOG(WARNING) << "Publish txn failed tablet:" << tablet->tablet_id() << " version:" << task.version
+                                 << " partition:" << task.partition_id << " txn_id: " << task.txn_id
+                                 << " rowset:" << task.rowset->rowset_id();
+                } else {
+                    LOG(INFO) << "Publish txn success tablet:" << tablet->tablet_id() << " version:" << task.version
+                              << " partition:" << task.partition_id << " txn_id: " << task.txn_id
+                              << " rowset:" << task.rowset->rowset_id();
+                }
+                task.version = tablet->max_continuous_version_from_beginning().second;
+            });
+            if (st.is_service_unavailable()) {
+                int64_t retry_sleep_ms = 50 * retry_time;
+                LOG(WARNING) << "publish version threadpool is busy, retry in  " << retry_sleep_ms
+                             << "ms. txn_id: " << transaction_id << ", tablet:" << tablet_tasks[i].tablet_id;
+                // In general, publish version is fast. A small sleep is needed here.
+                SleepFor(MonoDelta::FromMilliseconds(retry_sleep_ms));
+                continue;
+            } else {
+                break;
+            }
+        }
+        if (!st.ok()) {
+            tablet_tasks[i].st = std::move(st);
+        }
+    }
+    threadpool.wait();
+
+    Status st;
+    finish_task.__isset.tablet_versions = true;
+    auto& error_tablet_ids = finish_task.error_tablet_ids;
+    auto& tablet_versions = finish_task.tablet_versions;
+    tablet_versions.reserve(tablet_tasks.size());
+    for (size_t i = 0; i < tablet_tasks.size(); i++) {
+        auto& task = tablet_tasks[i];
+        if (!task.st.ok()) {
+            error_tablet_ids.push_back(task.tablet_id);
+            if (st.ok()) {
+                st = task.st;
+            }
+        }
+        if (task.version > 0) {
```

Comment:
If publish failed, does it also set the version of pair?

Reply:
yes, it will always report version